src/sapi.rs

```diff
@@ -246,6 +246,71 @@ impl SpVoice {
         self.set_rate(rate)
     }
 
+    pub fn get_voice_name(&mut self, token: Option<Speech::ISpObjectToken>) -> String {
+        unsafe {
+            token
+                .and_then(|t| t.OpenKey(w!("Attributes")).ok())
+                .and_then(|k| k.GetStringValue(w!("name")).ok())
+                .and_then(|s| s.to_string().ok())
+                .unwrap_or("unknown".to_string())
+        }
+    }
+
+    pub fn available_voices(&mut self) -> Vec<Speech::ISpObjectToken> {
+        let mut voices = vec![];
+        unsafe {
+            let category: Speech::ISpObjectTokenCategory =
+                syscom::CoCreateInstance(&Speech::SpObjectTokenCategory, None, syscom::CLSCTX_ALL)
+                    .expect("create voice category");
+            category
+                .SetId(
+                    w!(r"HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Speech\Voices"),
+                    false,
+                )
+                .expect("set voice category id");
+
+            let token_enum = category.EnumTokens(w!(""), w!("")).expect("get voice list");
+            loop {
+                let mut token = MaybeUninit::uninit();
+                token_enum
+                    .Next(1, token.as_mut_ptr(), None)
+                    .expect("iterate voices");
+                match token.assume_init() {
+                    Some(t) => {
+                        voices.push(t);
+                    }
+                    None => {
+                        break;
+                    }
+                }
+            }
+        }
+        voices
+    }
+
+    pub fn available_voice_names(&mut self) -> Vec<String> {
+        self.available_voices()
+            .iter()
+            .map(|t| self.get_voice_name(Some(t.clone())))
+            .collect::<Vec<_>>()
+    }
+
+    pub fn set_voice(&mut self, token: Speech::ISpObjectToken) {
+        unsafe { self.voice.SetVoice(&token).expect("set voice") }
+    }
+
+    pub fn set_voice_by_name(&mut self, voice_name: String) -> String {
+        if let Some(t) = self
+            .available_voices()
+            .iter()
+            .find(|&t| voice_name == self.get_voice_name(Some(t.clone())))
+        {
+            self.set_voice(t.clone());
+        }
+        // Return name of voice now in use
+        self.get_voice_name(unsafe { self.voice.GetVoice().ok() })
```

Comment:
I was matching the behavior of `set_rate`, which calls `get_rate` as the return value. It's useful when calling from `reload_settings`, which gets and sets values at the same time.

I've updated so that `get_voice_name` doesn't take an `Option`, though it doesn't really seem to simplify anything.

Reply:
A great example of why code reviews should be a conversation, often the person you just submitted the code understands its advantages better than the person with more experience.